src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -0,0 +1,282 @@
+            }
+
+            public void AnalyzeMethod(SymbolAnalysisContext context)
+            {
+                IMethodSymbol method = (IMethodSymbol)context.Symbol;
+
+                DllImportData? dllImportData = method.GetDllImportData();
+                if (dllImportData is null)
+                {
+                    return;
+                }
+
+                if (dllImportData.SetLastError)
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, SetLastErrorTrue));
+                }
+
+                if (!method.MethodImplementationFlags().HasFlag(System.Reflection.MethodImplAttributes.PreserveSig))
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, HResultSwapping));
+                }
+
+                if (_lcidConversionAttribute is not null && method.GetAttributes(_lcidConversionAttribute).Any())
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, UsingLCIDConversionAttribute));
+                }
+
+                if (method.IsVararg)
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, VarargPInvokes));
+                }
+
+                AnalyzeMethodSignature(context.ReportDiagnostic, method);
+            }
+
+            private void AnalyzeMethodSignature(Action<Diagnostic> reportDiagnostic, IMethodSymbol method, ImmutableArray<Location> locationsOverride = default)
+            {
+                AnalyzeSignatureType(locationsOverride.IsDefaultOrEmpty ? method.Locations : locationsOverride, method.ReturnType);
+                foreach (var param in method.Parameters)
+                {
+                    var paramLocation = locationsOverride.IsDefaultOrEmpty ? param.Locations : locationsOverride;
+                    if (param.RefKind != RefKind.None)
+                    {
+                        reportDiagnostic(paramLocation.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, ByRefParameters));
+                    }
+                    AnalyzeSignatureType(paramLocation, param.Type);
+                }
+
+                void AnalyzeSignatureType(ImmutableArray<Location> locations, ITypeSymbol type)
+                {
+                    if (!type.IsUnmanagedType)
+                    {
+                        reportDiagnostic(locations.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, ManagedParameterOrReturnTypes));
+                    }
+
+                    if (type.IsValueType && TypeIsAutoLayoutOrContainsAutoLayout(type))
+                    {
+                        reportDiagnostic(locations.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, ByRefParameters));
```

Comment:
So, use of localized strings as string format arguments for diagnostic messages is not recommended. This can turn out to generate incorrect localized strings for various locales. Instead, you should use different localized strings for each of these combinations of `FeatureUnsupportedWhenRuntimeMarshallingDisabled` diagnostics. This would mean you need to create different diagnostic descriptors for each of these district messages, all of them using the same diagnostic ID, but just a different message, which is fine, we do it at bunch of places in the repo. For example, see this for quite a descriptor explosion: https://github.com/dotnet/roslyn-analyzers/blob/577deb8e3365755b7d047c4b411d4a86f852be59/src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs#L53-L211

We are forced to take this approach to ensure proper localized strings for diagnostic messages, no other alternatives really work well here.